clients/include/testing_reduction_batched.hpp

```diff
@@ -0,0 +1,201 @@
+        return;
+    }
+
+    host_vector<R> hr1(batch_count);
+    CHECK_HIP_ERROR(hr1.memcheck());
+    host_vector<R> hr2(batch_count);
+    CHECK_HIP_ERROR(hr2.memcheck());
+    host_vector<R> cpu_result(batch_count);
+    CHECK_HIP_ERROR(cpu_result.memcheck());
+    device_vector<R> dr(batch_count);
+    CHECK_HIP_ERROR(dr.memcheck());
+
+    host_batch_vector<T> hx(N, incx, batch_count);
+    CHECK_HIP_ERROR(hx.memcheck());
+    device_batch_vector<T> dx(N, incx, batch_count);
+    CHECK_HIP_ERROR(dx.memcheck());
+
+    //
+    // Initialize.
+    //
+    rocblas_init(hx, true);
+
+    //
+    // Copy data from host to device.
+    //
+    CHECK_HIP_ERROR(dx.transfer_from(hx));
+
+    double gpu_time_used, cpu_time_used;
+    if(arg.unit_check || arg.norm_check)
+    {
+        //
+        // GPU BLAS, rocblas_pointer_mode_host
+        //
+        {
+            CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
+            CHECK_ROCBLAS_ERROR((func(handle, N, dx.ptr_on_device(), incx, batch_count, hr1)));
+        }
+
+        //
+        // GPU BLAS, rocblas_pointer_mode_device
+        //
+        {
+            CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_device));
+            CHECK_ROCBLAS_ERROR((func(handle, N, dx.ptr_on_device(), incx, batch_count, dr)));
+            CHECK_HIP_ERROR(hr2.transfer_from(dr));
+        }
+
+        //
+        // CPU BLAS
+        //
+        {
+            cpu_time_used = get_time_us();
+            for(rocblas_int batch_index = 0; batch_index < batch_count; ++batch_index)
+            {
+                CBLAS_FUNC(N, hx[batch_index], incx, cpu_result + batch_index);
+                //
+                // make index 1 based as in Fortran BLAS, not 0 based as in CBLAS
+                //
+                *(cpu_result + batch_index) += 1;
+            }
+            cpu_time_used = get_time_us() - cpu_time_used;
+        }
+
+        if(arg.unit_check)
+        {
+            unit_check_general<R>(batch_count, 1, 1, cpu_result, hr1);
+            unit_check_general<R>(batch_count, 1, 1, cpu_result, hr2);
+        }
+
+        if(arg.norm_check)
+        {
+            rocblas_error_1 = 0.0;
+            rocblas_error_2 = 0.0;
+            for(rocblas_int batch_index = 0; batch_index < batch_count; ++batch_index)
+            {
+                double a1       = double(hr1[batch_index]);
+                double a2       = double(hr2[batch_index]);
+                double c        = double(cpu_result[batch_index]);
+                rocblas_error_1 = std::max(rocblas_error_1, std::abs((c - a1) / c));
+                rocblas_error_2 = std::max(rocblas_error_2, std::abs((c - a2) / c));
+            }
+        }
+    }
+
+    if(arg.timing)
+    {
+        int number_cold_calls = 2;
+        int number_hot_calls  = 100;
+        CHECK_ROCBLAS_ERROR(rocblas_set_pointer_mode(handle, rocblas_pointer_mode_host));
```

Comment:
We should check to see if we are timing for fetching host results always, or would the device mode be better?

Reply:
To see with Wasiq

Reply:
> 
> 
> We should check to see if we are timing for fetching host results always, or would the device mode be better?

Host mode is usually better, if any scalar arguments have to be used on the host. We can pass scalars from the host to the device in the kernel call after loading them on the host, but we cannot pass scalars from the device to host without using a synchronizing memory copy.

Pointer mode only has to do with scalars. It does not affect vectors, matrices, tensors, etc., which area always on the device memory.